text/3649-macros-named-capture-groups.md

```diff
@@ -0,0 +1,334 @@
+- Feature Name: `macros-named-capture-groups`
+- Start Date: 2024-05-28
+- RFC PR: [rust-lang/rfcs#3649](https://github.com/rust-lang/rfcs/pull/3649)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+It will now be possible to give names to capture (repetition) groups in macro
+patterns, which can then be referred to directly in the macro body and macro
+metavariable expressions.
+
+# Motivation
+[motivation]: #motivation
+
+Rust has no way to refer to capture groups directly, so it uses the variables
+they capture to refer to them indirectly. This leads to confusing or limited
+behavior in a few places:
+
+- Expansion with multiple capture groups is extremely limited. In many cases,
+  the ordering and nesting of different groups is restricted based on what can
+  be inferred by the contained variables, since the groups themselves are
+  ambiguous.
+- Metavariable expressions use an unintuitive format: syntax like 
+  `${count($var, n)}` is used to refer to the `n`th parent group of the
+  smallest group that captures `$var`. Referring to groups directly would be
+  more straightforward than using a proxy.
+- Repetition-related diagnostics are limited because the compiler has limited
+  ability to guess what a capture group _should_ refer to when a captured
+  groups and variables do not align correctly.
+- Repetition mismatch diagnostics like can only be emitted after the macro is
+  instantiated, rather than when the macro is written. (E.g. "meta-variable
+  `foo` repeats 2 times, but `bar` repeats 1 time")
+- As a result of the above, using repetition is somewhat fragile; small
+  adjustments can break working patterns with little indication of what exactly
+  is wrong. Reading code with multiple capture groups can also be confusing.
+
+It is expected that named capture groups will provide a way to remove ambiguity
+in expansion and metavariable expressions, as well as unblock diagnostics that
+do a better job of guiding the macro mental model.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Capture groups can now take a name by providing an identifier and `:` between
+the `$` and opening `(`. This group can then be referred to by name in the
+expansion:
+
+```rust
+macro_rules! foo {
+    ( $group1:( $a:ident ),+ ) => {
+        $group1( println!("{}", $a); )+
+    }
```

Comment:
when you write this I thought you are defining a metavariable `$group1` which matches a comma-separated list of `$a:ident`, and using `$group1` will expand to `foo,bar,baz,etc`.

but no the `$group1` is just a label :thinking: 

the rationale mentioned using `:` is to be "to be more consistent with existing fragment specifiers" but IMO similarity with fragment specifiers  *is exactly why this RFC should not use the current syntax*.

i prefer getting rid of `:` (my main source of confusion) and use a label syntax like

```rust
macro_rules! make_functions {
    ( 
        names: [ $'names $($name:ident),+ ],
//               ^~~~~~~~~
        $'greetings $( greeting: $greeting:literal, )?
    ) => {
        $'names $(
            fn $name() {
                println!("function {} called", stringify!($name));
                $'greetings $(println!("{}", $greeting) )?
            }
        )+
    }
}


macro_rules! innermost1 {
    ( $'outer_rep $($a:ident: $'inner_rep $($b:literal),* );+ ) => {
        [$'outer_rep $( $'inner_rep $( ${index($'outer_rep)}, )* )+]
    };
}
```

Reply:
Metavariables matching some parts of the input are an interesting idea in general.
If we also introduce "match exactly once repetitions", then we'll also be able to capture just arbitrary token streams.
It would be nice to also be able to bind a name to "repeated exactly one" groups aka just artibtrary substreams in the input.
```rust
// macro LHS, capture "exactly once" repetition
$my_tokens:(a b $var c d)①

// macro RHS, reemit (although finding a good way to substitute variables is an interesting problem)
$my_tokens(value_for_var)
```

Reply:
Thanks for the ideas - I updated the default syntax to be `$group1(...)` (no colon) and added the label syntax as a possibility. 

>  It would be nice to also be able to bind a name to "repeated exactly one" groups aka just artibtrary substreams in the input.

Is there a standard kleene for a single capture, or could we get by with no kleene at all? I could probably include a  proposal for single captures in this RFC, to be accepted as part of it or rejected separately.

Reply:
In regexes, a "capture exactly once" group is just `(...)`.

Reply:
After writing some more examples, it seems like "capture once" and "emit the entire capture" would be pretty useful, so I just included them as part of the proposal. Especially for code that just needs to be validated and passed elsewhere, or matching and reemitting optional keywords/tts like `mut`, `&` or `pub(crate)` (just allowing capture groups that don't contain metavars is enough for that, but `$pub_crate` in the expansion is nicer to read than `$pub_crate(pub(crate))`).

Reply:
One advantage of the reemitting is that it will use spans from the input, instead of spans from the macro body.

One example I recently observed in https://github.com/rust-lang/rust/pull/119412 is the `mir!` macro in standard library.
It will match on something like `let $expr = $expr;` and then reemit `$expr = $expr;` as an assignment expression.
If `=` is emitted manually from the macro it will get the macro span and may mess up the combined span of the assignment expression.
If `=` is taken from the input, then the expression will get a precise span from the input, which is good for any kind of DSLs.